File: lang/c/test/fun/vac/data_structures/CircularlyLinkedListTest.c
#include <fun/vac/data_structures/CircularlyLinkedList.h>
#include <fun/vac/util/TestRunner.h>

bool testCircularlyLinkedList(void) {
    size_t size = 8192;

    CircularlyLinkedList *list = CircularlyLinkedList_new();

    for (size_t i = 1; i <= size; i++) {
        CircularlyLinkedList_append(list, (int) i);
    }

    if (CircularlyLinkedList_size(list) != size) {
        return false;
    }

    for (size_t i = 0; i < size; i++) {
        if ((size_t) CircularlyLinkedList_get(list, i) != i + 1) {
            return false;
        }
    }

    for (size_t i = 0; i < size; i++) {
        CircularlyLinkedList_set(list, i, size - i);
    }

    for (size_t i = 0; i < size; i++) {
        if ((size_t) CircularlyLinkedList_get(list, i) != size - i) {
            return false;
        }
    }

    int x, y;

    for (size_t i = 0, j = size - 1; i < j; i++, j--) {
        x = CircularlyLinkedList_get(list, i);
        y = CircularlyLinkedList_get(list, j);

        CircularlyLinkedList_remove(list, i);
        CircularlyLinkedList_insert(list, i, y);
        CircularlyLinkedList_remove(list, j);
        CircularlyLinkedList_insert(list, j, x);
    }

    for (size_t i = 0; i < size; i++) {
        if ((size_t) CircularlyLinkedList_get(list, i) != i + 1) {
            return false;
        }
    }

    for (size_t i = size; i >= 1; i--) {
        if ((size_t) CircularlyLinkedList_back(list) != i) {
            return false;
        } else {
            CircularlyLinkedList_eject(list);
        }
    }

    if (!CircularlyLinkedList_isEmpty(list)) {
        return false;
    }

    return true;
}

int main(void) {
    TestRunner_parseTest(testCircularlyLinkedList());
}

File: lang/ruby/test/vac.fun/data_structures/CircularlyLinkedListTest.rb
require "vac.fun/data_structures/CircularlyLinkedList"
require "vac.fun/util/TestRunner"

def testCircularlyLinkedList()
    size = 8192

    list = CircularlyLinkedList::CircularlyLinkedList.new()

    1.upto(size) do |i|
        list.append(i)
    end

    if list.size() != size
        return false
    end

    0.upto(size - 1) do |i|
        if list.get(i) != i + 1
            return false
        end
    end

    0.upto(size - 1) do |i|
        list.set(i, size - i)
    end

    0.upto(size - 1) do |i|
        if list.get(i) != size - i
            return false
        end
    end

    x = 0
    y = 0

    i, j = 0, size - 1
    while i < j
        x = list.get(i)
        y = list.get(j)

        list.remove(i)
        list.insert(i, y)
        list.remove(j)
        list.insert(j, x)
        i, j = i + 1, j - 1
    end

    0.upto(size - 1) do |i|
        if list.get(i) != i + 1
            return false
        end
    end

    size.downto(1) do |i|
        if list.back() != i
            return false
        end
        list.eject()
    end

    list.isEmpty()
end

if __FILE__ == $PROGRAM_NAME
    TestRunner.parseTest(testCircularlyLinkedList())
end
